src/transforms/lua/v2/interop/metric.rs

```diff
@@ -0,0 +1,539 @@
+                aggregated_summary.set("values", values)?;
+                aggregated_summary.set("count", count)?;
+                aggregated_summary.set("sum", sum)?;
+                tbl.set("aggregated_summary", aggregated_summary)?;
+            }
+        }
+
+        Ok(LuaValue::Table(tbl))
+    }
+}
+
+impl<'a> FromLua<'a> for Metric {
+    fn from_lua(value: LuaValue<'a>, _: LuaContext<'a>) -> LuaResult<Self> {
+        let table = match value {
+            LuaValue::Table(table) => table,
+            _ => {
+                return Err(LuaError::FromLuaConversionError {
+                    from: "",
+                    to: "Metric",
+                    message: Some("Metric should be a Lua table".to_string()),
+                })
+            }
+        };
+
+        let name: String = table.get("name")?;
+        let timestamp = table
+            .get::<_, Option<LuaTable>>("timestamp")?
+            .map(|t| table_to_timestamp(t))
+            .transpose()?;
+        let tags: Option<BTreeMap<String, String>> = table.get("tags")?;
+        let kind = table
+            .get::<_, Option<MetricKind>>("kind")?
+            .unwrap_or(MetricKind::Absolute);
+
+        let value = if let Some(counter) = table.get::<_, Option<LuaTable>>("counter")? {
+            MetricValue::Counter {
+                value: counter.get("value")?,
+            }
+        } else if let Some(gauge) = table.get::<_, Option<LuaTable>>("gauge")? {
+            MetricValue::Gauge {
+                value: gauge.get("value")?,
+            }
+        } else if let Some(set) = table.get::<_, Option<LuaTable>>("set")? {
+            MetricValue::Set {
+                values: set.get::<_, Vec<String>>("values")?.into_iter().collect(), // FIXME
+            }
+        } else if let Some(distribution) = table.get::<_, Option<LuaTable>>("distribution")? {
+            MetricValue::Distribution {
+                values: distribution.get("values")?,
+                sample_rates: distribution.get("sample_rates")?,
+            }
+        } else if let Some(aggregated_histogram) =
+            table.get::<_, Option<LuaTable>>("aggregated_histogram")?
+        {
+            let counts: Vec<u32> = aggregated_histogram.get("counts")?;
+            let count = counts.iter().sum();
+            MetricValue::AggregatedHistogram {
+                buckets: aggregated_histogram.get("buckets")?,
+                counts,
+                count,
+                sum: aggregated_histogram.get("sum")?,
+            }
+        } else if let Some(aggregated_summary) =
+            table.get::<_, Option<LuaTable>>("aggregated_summary")?
+        {
+            MetricValue::AggregatedSummary {
+                quantiles: aggregated_summary.get("quantiles")?,
+                values: aggregated_summary.get("values")?,
+                count: aggregated_summary.get("count")?,
+                sum: aggregated_summary.get("sum")?,
+            }
+        } else {
```

Comment:
It's unfortunate we can't make this a `match` statement to avoid this breaking accidentally in future metric type additions.

Reply:
I considered devising a serde-style `derive` macro which would implement `FromLua` for arbitrary enums, but decided against in the initial implementation because I thought that it might add too much complexity, and also it is not clear what to do with `DateTime<Utc>` and `Bytes` because both of these types are external, so `FromLua` cannot be implemented for them and it has to be worked around somehow.

But doing this this could have solved the issue of possible breaking of this in case if new metric type is added.